src/decimalParser.ts

```diff
@@ -0,0 +1,336 @@
+import {
+  StateConverterOptions,
+  StateConverterOptionsWithContext
+} from "./converter";
+
+type TOKEN_MINUS = "-";
+type TOKEN_DECIMAL_SEPARATOR = ".";
+type TOKEN_THOUSAND_SEPARATOR = ",";
+type TOKEN_WHITESPACE = " ";
+type TOKEN_DIGIT = "0";
+
+const TOKEN_MINUS: TOKEN_MINUS = "-";
+const TOKEN_DECIMAL_SEPARATOR: TOKEN_DECIMAL_SEPARATOR = ".";
+const TOKEN_THOUSAND_SEPARATOR: TOKEN_THOUSAND_SEPARATOR = ",";
+const TOKEN_WHITESPACE: TOKEN_WHITESPACE = " ";
+const TOKEN_DIGIT: TOKEN_DIGIT = "0";
+
+type TokenType =
+  | TOKEN_MINUS
+  | TOKEN_DECIMAL_SEPARATOR
+  | TOKEN_THOUSAND_SEPARATOR
+  | TOKEN_WHITESPACE
+  | TOKEN_DIGIT;
+
+const DIGIT = new RegExp("\\d");
+const WHITESPACE = new RegExp("\\s");
+
+function isDigit(c: string): boolean {
+  return DIGIT.test(c);
+}
+
+function isWhitespace(c: string): boolean {
+  return WHITESPACE.test(c);
+}
+
+export type DecimalOptions = {
+  maxWholeDigits: number;
+  decimalPlaces: number;
+  addZeroes: boolean;
+  allowNegative: boolean;
+};
+
+type TokenOptions = {
+  decimalSeparator: string;
+  thousandSeparator: string;
+  renderThousands: boolean;
+};
+
+type Options = DecimalOptions & TokenOptions;
+
+type Accept = (tokenType: TokenType) => boolean;
+type Expect = (tokenType: TokenType) => boolean;
+type NextToken = () => void;
+
+class Token {
+  constructor(public type: TokenType, public value: string) {}
+}
+
+function thousands(wholeDigits: string, thousandSeparator: string): string {
+  return wholeDigits.replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator);
+}
+
+function trimAndZeroes(
```

Comment:
This function name is a bit awkward IMO. Essentially this corrects the number of decimals shown in the field. If there aren't enough, it adds zeroes (when turned on); if there are too many, it truncates. I don't have any particularly great suggestions for a better name though. Best I can think of is correctNumberOfDecimals.